### ISSUE TITLE
ability to embed slint component from different source

### ISSUE THREAD

**Issue by ogoffart** [slint-ui/slint#2390]
Use case:
 - Slint Editor 
 - Editor that allow to place Slint Component
 - Dynamic widget from a model
 - Dynamic load .slint files at runtime

We would introduce a new `component` property type, and a `Embed` builtin element:

```slint
component Editor {
   in property <component> foo;
   // ...
        Embed { component: foo; }
}
```

The generated code for the setter would be something like

```rust
fn set_foo(_ : Option<impl slint::ComponentHandle>);
```

We should try to create the window adapter lazily because the `Embed` will set it when the sub-tree is visited (by its surrounding WindowAdapter)

The geometry constraints are forwarded from the embedded component to the `Embedd` element.
There will be a two way binding for the width and height of the Embedd element, and its component.

If a component is used several times, we need to do either:
 1. exit the event loop with a specific error saying this hapenned
 2. just print a error in stderr
The Embed should know its source location at runtime to give the developer some idea on where the error came from.

**Comment by ogoffart** [comment]
The question is whether a `component` is the right property type name. Would it be confusing?

I'm sure people will try things like this

```
// This is a component, can i put it in a property?
component Foo { ... }

component Bar {
   foo1 := Foo { }
   // not allowed  (foo1 is an element)
   property <component> p1: foo1;  
   // not allowed. We actually need an instance of a component
   property <component> p2: Foo;
   // Grammar don't currently support that. Should it? what is the parent then?  How can i even refer to other id in the current document, that won't work without a garbage collector
   property <component> p3: Foo { width: parent.width; height: foo1.height; };
   // If we allow something like the previous line that means that Foo {} becomes an expression,
   // then the gramar should alow things like that
   property <component> p4: true ? Foo { ... } : SomethingElse { ... };
}
```

So i'm just afraid that the `component` type might be confusing and people will try to do things with it that is not what it is meant for.  Another name would be `component-instence`

**Comment by tronical** [comment]
You're right, `component` is not a good name for the type. `component-instance` is clearer, I agree. If `component-instance` is a weak pointer, should there be also a strong equivalent?

**Comment by ogoffart** [comment]
Back in the meeting, we also discussed the concept of component factory.  And in fact, i think we should revisit this option.
I don't have a name for it, but it is just like a model, but instead of creating a T, it creates a component.
So given this Slint code

```
component FooBar {
    in property <component-factory> my_component_model;
    // ...
         for xxx in 42: Rectangle {
              // ...
                  Embed {
                       component: my_component_model;
                  }
         }
}
```

in rust
```rust
let foobar = FooBar::new().unwrap();
// `component-factory` takes a `Rc<dyn slint::ComponentFactory>`
foobar.set_my_component_factory(Rc::new(MyComponentFactory::new()));


impl slint::ComponentFactory for MyComponentFactory {
    fn build_component(&self) -> ComponentHandle {
           let result = self.the_component_definition.create();
           // store so we can update it with property changes.
           self.current_component.borrow_mut() = result.as_weak()
           result
    }
}
```

**Comment by ogoffart** [comment]
> I think the ComponentHandle will need to have it's Inner type to be defined. So we end up with a generic function. Or we just return a ComponentWeak from the build_component function (or the corresponding VRc).

True, we may need to change ComponentHandle to no longer have an Inner, or create a new trait that can be used as a trait object.

> Instead of just adding a set_parent_node function, I want to wrap my component with one that basically just overrides the parent_node function (to return the right component/index) and forwards everything else. 

I'm thinking having a set_parent_node may be the simpler.  The wrapper option might be complicated, because the inner node of a repeated component will have a weak pointer to the non-wrapper as a parent.

**Comment by ogoffart** [comment]
> think the ComponentHandle will need to have it's Inner type to be defined

Only if we use `dyn` but anyway, we can't return a trait from a function in an object safe trait. So we'll have to do something like

```
/// Wrapper around something that implements ComponentHandle.
/// Can be constructed with `ComponentInstance::from(my_component)` where my_component immplements
/// the ComponentHandle trait. (eg: slint_interpreter::ComponentInstance or the type generated from 
/// the slint code.
///
/// FIXME: naming?
pub struct ComponentInstance(crate::component::ComponentRc);
impl<T: ComponentHandle> From<T> for ComponentInstance
where
    T::Inner: vtable::HasStaticVTable<ComponentVTable> + 'static,
{
    fn from(value: T) -> Self {
        Self(vtable::VRc::into_dyn(value.as_weak().inner().upgrade().unwrap()))
    }
}


/// The type `component-factory`  in slint maps to a Rc<dyn ComponentFactory>
trait ComponentFactory {
    fn build(&self) -> ComponentInstance;
}
```

This is going to make C++ a bit tricky because `Rc<dyn ComponentFactory>` is not ffi safe, so we'll have to use the same trick as with `WindowAdapterRcOpaque`. That adds one memory allocation indirection, but the alternative is to use  a `vtable::VRc<ComponentFactoryVTable>`  or a wrapper around it, and that sucks.

**Comment by ales-tsurko** [comment]
I see [`ComponentFactory`](https://slint.dev/releases/1.2.0/docs/rust/slint/struct.ComponentFactory) in the rust docs. Can I somehow use it to add components at the runtime?

**Comment by bryce-happel-walton** [comment]
I assume this will be similar to Qt's [Loader](https://doc.qt.io/qt-6/qml-qtquick-loader.html). I see that the [ComponentFactory](https://releases.slint.dev/1.2.0/docs/rust/slint/struct.componentfactory) is no longer in the latest Rust documentation. Has this feature been scrapped?

**Comment by Caellian** [comment]
I don't think any of suggested default initialization methods need to be supported - the main purpose of these property declarations would be defining them from native code or passing them in from the parent (which also got them from somewhere else).

`Embed` needs `reserve-width` and `reserve-height` properties for layouting purposes - like how `opacity: 0` doesn't affect layout while `visible: false` does.

You can swap out the component with the default by checking whether the property is `null` - so does `Embed` gracefully handle `null` values? If so, default initialization can be completely skipped which avoids complicating `property` syntax or any other alternative; at least for MVP version.

Functions aren't values and have `callback` syntax, so you can add alternative syntax like:

```slint
component MenuOption {
  in property <string> name: "Option Item";
  default-interact := Button { text: "Do Action" }
  
  // Added a default for example purposes, not really needed
  in instance <SpinBox|Button> interactive: default-interact;
  
  // ... layout and styles specific to my library ...
  Text { text: name }
  Embed { component: interactive }
  Rectangle { visible: interactive == null } // Placeholder if not set
}
```

and that would both make it clear that it's an instance of a `Rectangle` and also allow constraints. `<T>` can either be optional, or maybe `<?>` would work given that there's no type for root component class exposed. Constraints might prove useful for component libraries as demonstrated in above example.

Reason I'm suggesting new syntax is that it opens up a path for incremental improvements - you can allow inline initialization later and simply provide a lint that `default-interact` can be inlined, and even that property with component-instance type should be used instead once each part is supported.

I think this is VERY important because it opens up the door for custom component rendering (at scale) - it _is_ possible atm, but I'd have to wrangle a bunch of global bindings to get it to work and that's really nasty (imagine making a UI with >20 of those).

EDIT: Rewrote parts so they sound more like input and less like demands. Added more details and thoughts.

**Comment by LeonMatthes** [comment]
Based on the discussion around named slots in #6258, @ogoffart and I came up with another possible, potentially simplified syntax:

```slint
component MyComponent {
  // Exact type TBD
  property <factory<Text>> factory;

  // single instance from a factory
  dropped-inst := factory {
    text: "Hello World";
  }

  // can be used as often as you want
  for i in [1, 2, 3]: factory {
    text: "{i}";
  }
}

export component Test inherits Window {
  property <bool> clicked: false;
  MyComponent {
    // @factory to enter a new "element" scope.
    component-factory: @factory(
      Text {
        // Factories could act almost like "closures"
        // that use weak references to anything they reference
        // but that is a stretch goal
        if root.thing: Rectangle {
          background: red;
        }
      }
    )
  }
}
```

Note: @ogoffart I just realized that this doesn't yet address the question of "default" value for `factory`.
What if the property is just not assigned?

**Comment by ogoffart** [comment]
A factory that is default generates an empty component that has the default value for the type of all properties.

I have the feeling that it would be benefitial to have some syntax to differentiate the factory.

```slint
component Xxx {
   in property <component-factory<SomeInterface>> my-interface;
   xyz := @embed my-interface { 
      some-property
   }
   
   // can be any expression
   @embed MyGlobal.some-factory {}
   @embed root.some-callback(34) {}
   callback some-callback(int) -> component-factory;
```

**Comment by LeonMatthes** [comment]
@eira-fransham @ogoffart and I just had another chat about slots #10738 and element refernces in properties.
Especially regarding the types involved around this we ended up with an interesting conclusion for factories.

What if a component factory is just a normal callback/function that returns an element?
The idea is to differentiate between "weak" references (or IDs) of components, that could be used in e.g. properties, and owned references, which is what slots use.

Then the syntax for creating a factory would just be a normal function that returns an owned element.
It would not need a special type at all.

e.g.:

```slint
component MyComponent {
  // Exact type TBD
  callback factory() -> Text;

  // single instance from a factory
  dropped-inst := factory() {
    text: "Hello World";
  }

  // can be used as often as you want
  for i in [1, 2, 3]: factory() {
    text: "{i}";
  }
}

export component Test inherits Window {
  property <bool> clicked: false;

  MyComponent {
    factory => () {
        // @element just creates an owned element.
        @element(
            Text {
              if root.clicked: Rectangle {
                background: red;
              }
            }
          )
     }
  }
}
```

Open Question: what to do about the factory arguments? When is a factory re-evaluated?

**Comment by eira-fransham** [comment]
I think that the strong pointers (which cannot be stored across function/callback invocations) and weak pointers (which can, but do not prevent the component from being deallocated) should have different names in the code, even though with the original design as it described to me, the type was unambiguous based on usage. @LeonMatthes mentioned that there should be a distinction in the type system between "owned" and "strong" references, with owned references being required for operations such as parenting. Strong and owned references should have precisely identical semantics, but different invariants (similar to `str` vs `[u8]`). In Rust we can think of it being something like this:

```rust
// Does _not_ implement `Clone`.
struct OwnedRc<T>(Rc<T>);

impl From<OwnedRc<T>> for Rc<T> {
    fn from(value: OwnedRc<T>) -> Self {
        value.0.clone()
    }
}

impl<T> TryFrom<Rc<T>> for OwnedRc<T> {
    type Error = ..;

    fn try_from(value: Rc<T>) -> Result<Self, Self::Error> {
        if Rc::strong_count(&value) == 1 {
            Ok(Self(value))
        } else {
            Err(..)
        }
    }
}
```

With the reparenting operations requiring an `OwnedRc<T>`, and `@component(..)` returning an `OwnedRc<T>`.

My proposed syntax for owned, strong and weak pointer types in Slint would be:

- Owned: just the component name (e.g. `Rectangle`)
- Strong: `#Rectangle`, by analogy to `#id` in HTML
- Weak: `#Rectangle?`, and when the pointee is dropped the property is set to `none` with the relevant `changed` handlers etc being called

The latter might be too magical, but I think it should be relatively easy to explain to a user "a property that refers to a component needs to be optional, so it can be set to `none` if the component is deleted". That means that strong and weak references are, essentially, the same type as far as Slint devs are concerned.

**Comment by Jnesselr** [comment]
Throwing in our use case in case it's useful for the design discussion. We've got a tab strip in an app where each tab is a different component, and we want each tab's view created and owned from the Rust side: instantiate it, hold the strong handle there, wire callbacks from Rust, then feed a clone of the handle into a ComponentFactory when the tab becomes active. The container / factory side of that works the way we'd want — the same handle can be swapped out and back, state intact. We did look at the TabWidget, but it explicitly disallows having different types of tabs and doesn't match the UI we're going for either.

The thing that hurts is that each exported tab component still has to inherit Window. We'd rather have them Rectangle-rooted. The theme/palette globals we set on the host root don't propagate into the embedded child (each Window root seems to instantiate its own copy of the globals), so we end up re-broadcasting theme changes through Rust. And the Window root brings along title, default size, MenuBar, WindowAdapter, none of which mean anything once the component is embedded.

So really the win for us here would be being able to export non-Window components for embedding. The factory / container mechanism itself is fine.
